Announce Rust 1.89.0 (#1666)
rust-lang/blog.rust-lang.org: content/Rust-1.89.0.md

Because the input and output lifetimes belong to different groups, the lint will warn about this function, reducing confusion about when a value has a meaningful lifetime that isn't visually obvious.

The `mismatched_lifetime_syntaxes` lint supersedes the `elided_named_lifetimes` lint, which did something similar for named lifetimes specifically.

Future work on the `elided_lifetimes_in_paths` lint intends to split it into more focused sub-lints with an eye to warning about a subset of them eventually.

[elision]: https://doc.rust-lang.org/1.89/book/ch10-03-lifetime-syntax.html#lifetime-elision
[elided_lifetime_in_path]: https://github.com/rust-lang/rust/pull/46254
[2018-by-default]: https://github.com/rust-lang/rust/issues/54910
[bevy]: https://github.com/rust-lang/rust/issues/131725

### More x86 target features

The `target_feature` attribute now supports the `sha512`, `sm3`, `sm4`, `kl` and `widekl` target features on x86. Additionally a number of `avx512` intrinsics and target features are also supported on x86:

```rust
#[target_feature(enable = "avx512bw")]
pub fn cool_simd_code(/* .. */) -> /* ... */ {
    /* ... */
}

```

> ### Platform Support

Reviewer: https://github.com/rust-lang/blog.rust-lang.org/pull/1666#issuecomment-3129049497

Future work on the `elided_lifetimes_in_paths` lint intends to split it into more focused sub-lints with an eye to warning about a subset of them eventually.

[elision]: https://doc.rust-lang.org/1.89/book/ch10-03-lifetime-syntax.html#lifetime-elision
[elided_lifetime_in_path]: https://github.com/rust-lang/rust/pull/46254
[2018-by-default]: https://github.com/rust-lang/rust/issues/54910
[bevy]: https://github.com/rust-lang/rust/issues/131725

### More x86 target features

The `target_feature` attribute now supports the `sha512`, `sm3`, `sm4`, `kl` and `widekl` target features on x86. Additionally a number of `avx512` intrinsics and target features are also supported on x86:

```rust
#[target_feature(enable = "avx512bw")]
pub fn cool_simd_code(/* .. */) -> /* ... */ {
    /* ... */
}

```

### Cross-compiled doctests

Doctests will now be tested when running `cargo test --doc --target other_target`, this may result in some amount of breakage due to would-be-failing doctests now being tested.

Failing tests can be disabled by annotating the doctest with `ignore-<target>`:
```rust
/// ```ignore-x86_64
/// panic!("something")
/// ```

> /// pub fn my_function() { }

Author: ```suggestion
pub fn my_function() { }
```

The `mismatched_lifetime_syntaxes` lint supersedes the `elided_named_lifetimes` lint, which did something similar for named lifetimes specifically.

Future work on the `elided_lifetimes_in_paths` lint intends to split it into more focused sub-lints with an eye to warning about a subset of them eventually.

[elision]: https://doc.rust-lang.org/1.89/book/ch10-03-lifetime-syntax.html#lifetime-elision
[elided_lifetime_in_path]: https://github.com/rust-lang/rust/pull/46254
[2018-by-default]: https://github.com/rust-lang/rust/issues/54910
[bevy]: https://github.com/rust-lang/rust/issues/131725

### More x86 target features

The `target_feature` attribute now supports the `sha512`, `sm3`, `sm4`, `kl` and `widekl` target features on x86. Additionally a number of `avx512` intrinsics and target features are also supported on x86:

```rust
#[target_feature(enable = "avx512bw")]
pub fn cool_simd_code(/* .. */) -> /* ... */ {
    /* ... */
}

```

### Cross-compiled doctests

Doctests will now be tested when running `cargo test --doc --target other_target`, this may result in some amount of breakage due to would-be-failing doctests now being tested.

> Failing tests can be disabled by annotating the doctest with `ignore-<target>`:

Reviewer 2: ```suggestion
Failing tests can be disabled by annotating the doctest with `ignore-<target>` ([docs](https://doc.rust-lang.org/stable/rustdoc/write-documentation/documentation-tests.html#ignoring-targets)):
```

```rust
/// ```ignore-x86_64
/// panic!("something")
/// ```
pub fn my_function() { }
``` 

### `i128` and `u128` in `extern "C"` functions

`i128` and `u128` no longer trigger the `improper_ctypes_definitions` lint, meaning these types may be used in `extern "C"` functions without warning. This comes with some caveats:

* The Rust types are ABI- and layout-compatible with (unsigned) `__int128` in C when the type is available.
* On platforms where `__int128` is not available, `i128` and `u128` do not necessarily align with any C type.
* `i128` is _not_ necessarily compatible with `_BitInt(128)` on any platform, because `_BitInt(128)` and `__int128` may not have the same ABI (as is the case on x86-64).

This is the last bit of follow up to the layout changes from last year: https://blog.rust-lang.org/2024/03/30/i128-layout-update/.

### Demoting `x86_64-apple-darwin` to Tier 2 with host tools

> GitHub will soon [discontinue][gha-sunset] providing free macOS x86\_64 runners for public repositories. Apple has also announced their [plans][apple] for discontinuing support for the x86\_86 architecture.

Reviewer 2: ```suggestion
GitHub will soon [discontinue][gha-sunset] providing free macOS x86\_64 runners for public repositories. Apple has also announced their [plans][apple] for discontinuing support for the x86\_64 architecture.
```

> In accordance with these changes, the Rust project is in the [process of demoting the `x86_64-apple-darwin` target][rfc] from Tier 1 to Tier 2 with host tools. This means that the target, including tools like `rustc` and `cargo`, will be guaranteed to build but is not guaranteed to pass our automated test suite.

Reviewer 2: ```suggestion
In accordance with these changes, the Rust project is in the [process of demoting the `x86_64-apple-darwin` target][rfc] from [Tier 1 with host tools](https://doc.rust-lang.org/stable/rustc/platform-support.html#tier-1-with-host-tools) to [Tier 2 with host tools](https://doc.rust-lang.org/stable/rustc/platform-support.html#tier-2-with-host-tools). This means that the target, including tools like `rustc` and `cargo`, will be guaranteed to build but is not guaranteed to pass our automated test suite.
```

We expect that the RFC for the demotion to Tier 2 with host tools will be accepted between the releases of Rust 1.89 and 1.90, which means that Rust 1.89 will be the last release of Rust where `x86_64-apple-darwin` is a Tier 1 target.

> For users, nothing will change. Builds of both the standard library and the compiler will still be distributed by the Rust Project for use via `rustup` or alternative installation methods.

Reviewer 2: ```suggestion
For users, this change will not immediately cause impact. Builds of both the standard library and the compiler will still be distributed by the Rust Project for use via `rustup` or alternative installation methods while the target remains at Tier 2.
```

> For users, this change will not immediately cause impact. Builds of both the standard library and the compiler will still be distributed by the Rust Project for use via `rustup` or alternative installation methods while the target remains at Tier 2.

Reviewer 3: @traviscross I don't agree with this specific wording as now it sounds like *this change will eventually cause impact*, which is not the case to my knowledge. What specific event or impact are you trying to suggest at?

Reviewer 4: (I didn't write that text.)

That said, the change will eventually have effect, at least probabilistically, due to our lower standards and expectations for how changes and decisions affect a Tier 2 target.

Reviewer 3: > (I didn't write that text.)

> I think it's me you meant to ping?

Ah yes, *mea culpa*, apologies.

The intent of my original wording is that people don't have to anything different with respect to getting the target (e.g. interactions with `rustup`). 

I do not think *this* post should imply or assume anything about tier 3, because that's not the change occurring. If there's a demotion to tier 3, we already have an obligation to report that in similar ways (release notes, a blog post), so we don't need to address that possibility here.

If we are worried about communicating reduced test coverage, we should be explicit about that. The current wording is too nebulous and will easily lead to panicked people asking "what do you mean?". 

```suggestion
For users, this change will not immediately cause impact. Builds of both the standard library and the compiler will still be distributed by the Rust Project for use via `rustup` or alternative installation methods while the target remains at Tier 2. Over time, it's likely that reduced test coverage for this target will cause things to break or fall out of compatibility with no further announcements.
```

If you have a previous version of Rust installed via `rustup`, you can get 1.89.0 with:

```console
$ rustup update stable
```

If you don't have it already, you can [get `rustup`](https://www.rust-lang.org/install.html) from the appropriate page on our website, and check out the [detailed release notes for 1.89.0](https://doc.rust-lang.org/stable/releases.html#version-1890-2025-08-07).

If you'd like to help us out by testing future releases, you might consider updating locally to use the beta channel (`rustup default beta`) or the nightly channel (`rustup default nightly`). Please [report](https://github.com/rust-lang/rust/issues/new/choose) any bugs you might come across!

## What's in 1.89.0 stable

### Explicitly inferred arguments to const generics

Rust now supports `_` as an argument to const generic parameters, inferring the value from surrounding context:

```rust
pub fn make_bitset<const LEN: usize>() -> [bool; LEN] {
  [false; _]
}
```

Similar to the rules for when `_` is permitted as a type, `_` is not permitted as an argument to const generics when in a signature:

```rust
// This is not allowed
pub fn make_bitset<const LEN: usize>() -> [bool; _] {
  [false; LEN]
}

> ```

Author: cc @rust-lang/project-const-generics

### Mismatched lifetimes syntax lint

> TC/LANG to write this

Reviewer 2: I think it's worth keeping, yes.

### More x86 target features

The `target_feature` attribute now supports the `sha512`, `sm3`, `sm4`, `kl` and `widekl` target features on x86. Additionally a number of `avx512` intrinsics and target features are also supported on x86:

```rust
#[target_feature(enable = "avx512bw")]
pub fn cool_simd_code(/* .. */) -> /* ... */ {
    /* ... */
}

> ```

Author: cc @Amanieu @sayantn :3 I don't know much about this

Reviewer 5: Also specify that all avx512 intrinsics in `std::arch::{x86,x86_64}` are stable now (same for the other target features)

> 

Reviewer 6: Maybe s/x86\_64 macOS/`x86_64-apple-darwin`/.

### Mismatched lifetimes syntax lint

> ### Mismatched lifetime syntaxes lint

Reviewer 3: And “lifetime syntaxes” matches the name of the line
